File: python/coinspark-test.py
import sys, os.path, random
from coinspark import *


def GetNextInputLine(inputSource):
	inputLine=inputSource.readline()
	if len(inputLine)==0: # empty string indicates end of file
		return None;
	
	inputLine=inputLine.rstrip("\n")
	
	hashPos=inputLine.find(' # ')
	if hashPos>=0:
		inputLine=inputLine[:hashPos]
	
	return inputLine
	
	
def GetNextInputLines(inputSource, countLines):
	inputLines=[]
	
	while (countLines>0):
		countLines-=1
		
		inputLine=GetNextInputLine(inputSource)
		if inputLine is None:
			return None
			
		inputLines.append(inputLine)
		
	return inputLines


def ProcessInputContents(inputSource):
	header=GetNextInputLine(inputSource)
	GetNextInputLine(inputSource) # discard blank line
	
	if header=='CoinSpark Address Tests Input':
		ProcessAddressTests(inputSource)
		
	elif header=='CoinSpark AssetRef Tests Input':
		ProcessAssetRefTests(inputSource)
		
	elif header=='CoinSpark Script Tests Input':
		ProcessScriptTests(inputSource)
		
	elif header=='CoinSpark AssetHash Tests Input':
		ProcessAssetHashTests(inputSource)
		
	elif header=='CoinSpark Genesis Tests Input':
		ProcessGenesisTests(inputSource)
		
	elif header=='CoinSpark Transfer Tests Input':
		ProcessTransferTests(inputSource)
		
	elif header=='CoinSpark MessageHash Tests Input':
		ProcessMessageHashTests(inputSource)
		
		
def ProcessAddressTests(inputSource):
	print("CoinSpark Address Tests Output\n")
	
	while True:
		inputLine=GetNextInputLine(inputSource)
		if inputLine is None:
			break
			
		address=CoinSparkAddress()
		
		if address.decode(inputLine):
			sys.stdout.write(address.toString())
		else:
			sys.exit("Failed to decode address: "+inputLine)
			
		encoded=address.encode()
		
		if encoded!=inputLine:
			sys.exit("Encode address mismatch: "+encoded+" should be "+inputLine)
			
		if not address.match(address):
			sys.exit("Failed to match address to itself!")
			

def ProcessAssetRefTests(inputSource):
	print("CoinSpark AssetRef Tests Output\n")
	
	while True:
		inputLine=GetNextInputLine(inputSource)
		if inputLine is None:
			break
	
		assetRef=CoinSparkAssetRef()
	
		if assetRef.decode(inputLine):
			sys.stdout.write(assetRef.toString())
		else:
			sys.exit("Failed to decode AssetRef: "+inputLine)
		
		encoded=assetRef.encode()
	
		if encoded!=inputLine:
			sys.exit("Encode AssetRef mismatch: "+encoded+" should be "+inputLine)
	
		if not assetRef.match(assetRef):
			sys.exit("Failed to match assetRef to itself!")


def ProcessScriptTests(inputSource):
	print("CoinSpark Script Tests Output\n");
	
	while True:
		inputLines=GetNextInputLines(inputSource, 4)
		if inputLines is None:
			break
			
		countInputs, countOutputs, scriptPubKeyHex, dummy = inputLines
		
		metadata=CoinSparkScriptToMetadata(scriptPubKeyHex, True)
		if metadata is None:
			sys.exit("Could not decode script metadata: "+scriptPubKeyHex);
			
		# Read in the different types of metadata
		
		genesis=CoinSparkGenesis()
		hasGenesis=genesis.decode(metadata)
		
		paymentRef=CoinSparkPaymentRef()
		hasPaymentRef=paymentRef.decode(metadata)
		
		transfers=CoinSparkTransferList()
		hasTransfers=transfers.decode(metadata, countInputs, countOutputs)
		
		message=CoinSparkMessage()
		hasMessage=message.decode(metadata, countOutputs)
		
		# Output the toString()s
		
		if hasGenesis:
			sys.stdout.write(genesis.toString())
			
		if hasPaymentRef:
			sys.stdout.write(paymentRef.toString())
			
		if hasTransfers:
			sys.stdout.write(transfers.toString())
			
		if hasMessage:
			sys.stdout.write(message.toString())
			
		# Re-encode
		
		testMetadata=''
		testMetadataMaxLen=len(metadata)
		nextMetadataMaxLen=testMetadataMaxLen
		
		encodeOrder=['genesis', 'paymentRef', 'transfers', 'message']
		
		for encodeField in encodeOrder:
			triedNextMetadata=False
			
			if encodeField=='genesis':
				if hasGenesis:
					nextMetadata=genesis.encode(nextMetadataMaxLen)
					triedNextMetadata=True
			
			elif encodeField=='paymentRef':
				if hasPaymentRef:
					nextMetadata=paymentRef.encode(nextMetadataMaxLen)
					triedNextMetadata=True
			
			elif encodeField=='transfers':
				if hasTransfers:
					nextMetadata=transfers.encode(countInputs, countOutputs, nextMetadataMaxLen)
					triedNextMetadata=True
					
			elif encodeField=='message':
				if hasMessage:
					nextMetadata=message.encode(countOutputs, nextMetadataMaxLen)
					triedNextMetadata=True
					
			if triedNextMetadata:
				if nextMetadata is None:
					sys.exit("Failed to reencode "+encodeField+" metadata!")
				
				if len(testMetadata):
					testMetadata=CoinSparkMetadataAppend(testMetadata, testMetadataMaxLen, nextMetadata)
					if testMetadata is None:
						sys.exit("Insufficient space to append "+encodeField+" metadata!")
			
				else:
					testMetadata=nextMetadata
				
				nextMetadataMaxLen=CoinSparkMetadataMaxAppendLen(testMetadata, testMetadataMaxLen)
				
		# Test other library functions while we are here
		
		if hasGenesis:
			if not genesis.match(genesis, True):
				sys.exit("Failed to match genesis to itself!")
				
			if genesis.calcHashLen(len(metadata))!=genesis.assetHashLen: # assumes that metadata only contains genesis
				sys.exit("Failed to calculate matching hash length!")
				
			testGenesis=CoinSparkGenesis()
			testGenesis.decode(metadata)
			
			rounding=random.randint(-1, 1)
			
			testGenesis.setQty(0, 0)
			testGenesis.setQty(genesis.getQty(), rounding)
			
			testGenesis.setChargeFlat(0, 0)
			testGenesis.setChargeFlat(genesis.getChargeFlat(), rounding)
			
			if not genesis.match(testGenesis, False):
				sys.exit("Mismatch on genesis rounding!")
				
		if hasPaymentRef:
			if not paymentRef.match(paymentRef):
				sys.exit("Failed to match paymentRef to itself!")
				
		if hasTransfers:
			if not transfers.match(transfers, True):
				sys.exit("Failed to strictly match transfers to itself!")
				
			if not transfers.match(transfers, False):
				sys.exit("Failed to leniently match transfers to itself!")
				
		if hasMessage:
			if not message.match(message, True):
				sys.exit("Failed to strictly match message to itself!")
				
			if not message.match(message, False):
				sys.exit("Failed to leniently match message to itself!")
				
			messageEncode=message.encode(countOutputs, len(metadata)) # encode on its own to check calcHashLen()
			
			if message.calcHashLen(countOutputs, len(messageEncode))!=message.hashLen:
				sys.exit("Failed to calculate matching message hash length!")
				
		# Compare to the original
		
		encoded=CoinSparkMetadataToScript(testMetadata, True)
		
		if encoded!=scriptPubKeyHex:
			sys.exit("Encode metadata mismatch: "+encoded+" should be "+scriptPubKeyHex)
		
		checkMetadata=CoinSparkScriptToMetadata(CoinSparkMetadataToScript(testMetadata, False), False)
		
		if checkMetadata!=testMetadata:
			sys.exit("Binary metadata to/from script mismatch!")
			


def ProcessAssetHashTests(inputSource):
	print("CoinSpark AssetHash Tests Output\n")
	
	while True:
		inputLines=GetNextInputLines(inputSource, 10)
		if inputLines is None:
			break
			
		name, issuer, description, units, issueDate, expiryDate, interestRate, multiple, contractContent, dummy = inputLines
		
		contractContent=contractContent.encode('utf-8')
		
		hash=CoinSparkCalcAssetHash(name, issuer, description, units, issueDate, expiryDate, interestRate, multiple, contractContent)
		
		print(binascii.hexlify(hash).upper().decode('utf-8'))


def ProcessGenesisTests(inputSource):
	print("CoinSpark Genesis Tests Output\n")
	
	while True:
		inputLines=GetNextInputLines(inputSource, 7)
		if inputLines is None:
			break
	
		# Break apart and decode the input lines
		
		firstSpentTxId, firstSpentVout, metadataHex, outputsSatoshisString, outputsRegularString, feeSatoshis, dummy = inputLines

		genesis=CoinSparkGenesis()
		if not genesis.decode(CoinSparkHexToRawString(metadataHex)):
			sys.exit("Failed to decode genesis metadata: "+metadataHex)
			
		outputsSatoshis=outputsSatoshisString.split(',')
		outputsRegular=[bool(int(outputRegular)) for outputRegular in outputsRegularString.split(',')]		
		countOutputs=len(outputsSatoshis)
	
		validFeeSatoshis=genesis.calcMinFee(outputsSatoshis, outputsRegular)
			
		# Perform the genesis calculation
		
		if float(feeSatoshis)>=validFeeSatoshis:
			outputBalances=genesis.apply(outputsRegular)
		else:
			outputBalances=[0] * countOutputs
			
		# Output the results
		
		print("%.0f # transaction fee satoshis to be valid" % validFeeSatoshis)
		print(",".join(["%.0f" % outputBalance for outputBalance in outputBalances])+" # units of the asset in each output")
		print(genesis.calcAssetURL(firstSpentTxId, firstSpentVout)+" # asset web page URL\n")
		
		
def ProcessTransferTests(inputSource):
	print("CoinSpark Transfer Tests Output\n")
	
	while True:
		inputLines=GetNextInputLines(inputSource, 8)
		if inputLines is None:
			break
	
		# Break apart and decode the input lines
		
		(genesisMetadataHex, assetRefString, transfersMetadataHex, inputBalancesString,
			outputsSatoshisString, outputsRegularString, feeSatoshis, dummy) = inputLines

		genesis=CoinSparkGenesis()
		if not genesis.decode(CoinSparkHexToRawString(genesisMetadataHex)):
			sys.exit("Failed to decode genesis metadata: "+genesisMetadataHex)
			
		assetRef=CoinSparkAssetRef()
		if not assetRef.decode(assetRefString):
			sys.exit("Failed to decode asset reference: assetRefString")
			
		inputBalances=inputBalancesString.split(',')
		outputsSatoshis=outputsSatoshisString.split(',')
		outputsRegular=[bool(int(outputRegular)) for outputRegular in outputsRegularString.split(',')]	
		
		countInputs=len(inputBalances)
		countOutputs=len(outputsSatoshis)
	
		transfers=CoinSparkTransferList()
		if not transfers.decode(CoinSparkHexToRawString(transfersMetadataHex), countInputs, countOutputs):
			sys.exit("Failed to decode transfers metadata: "+transfersMetadataHex)
		validFeeSatoshis=transfers.calcMinFee(countInputs, outputsSatoshis, outputsRegular)
			
		# Perform the transfer calculation and get default flags
		
		if float(feeSatoshis)>=validFeeSatoshis:
			outputBalances=transfers.apply(assetRef, genesis, inputBalances, outputsRegular)
		else:
			outputBalances=transfers.applyNone(assetRef, genesis, inputBalances, outputsRegular)
			
		outputsDefault=transfers.defaultOutputs(countInputs, outputsRegular)
		
		# Output the results
		
		print("%.0f # transaction fee satoshis to be valid" % validFeeSatoshis)
		print(",".join(["%.0f" % outputBalance for outputBalance in outputBalances])+" # units of this asset in each output")
		print(",".join(['1' if outputDefault else '0' for outputDefault in outputsDefault])+" # boolean flags whether each output is in a default route\n")
		
		# Test the net and gross calculations using the input balances as example net values
	
		for inputBalance in inputBalances:
			testGrossBalance=genesis.calcGross(inputBalance)
			testNetBalance=genesis.calcNet(testGrossBalance)
			
			if float(inputBalance)!=testNetBalance:
				sys.exit("Net to gross to net mismatch: %.0f -> %.0f -> %.0f!" % (float(inputBalance), testGrossBalance, testNetBalance))
		

def ProcessMessageHashTests(inputSource):
	print("CoinSpark MessageHash Tests Output\n")
	
	while True:
		inputLines=GetNextInputLines(inputSource, 2)
		if inputLines is None:
			break
			
		salt, countParts = inputLines

		salt=salt.encode('utf-8')
		countParts=int(countParts)
		
		inputLines=GetNextInputLines(inputSource, 3*countParts+1)
		if inputLines is None:
			break
		
		messageParts=[]
		while ((len(messageParts)<countParts) and (len(inputLines)>0)):
			messageParts.append({
				'mimeType': inputLines.pop(0),
				'fileName': inputLines.pop(0),
				'content': inputLines.pop(0).encode('utf-8')
			})
			
		hash=CoinSparkCalcMessageHash(salt, messageParts)
		
		print(binascii.hexlify(hash).upper().decode('utf-8'))


# Main entry point for code
	
if len(sys.argv)<2:
	sys.exit("No CoinSpark test input file was specified\n")

inputFileName=sys.argv[1]

if os.path.isfile(inputFileName):
	inputSource=open(inputFileName, 'r')
	ProcessInputContents(inputSource);
	inputSource.close()

else:
	sys.exit("The CoinSpark test input file was wrong\n")
	

File: php/coinspark-test.php
<?php

/*
 * CoinSpark 2.1 - PHP test suite
 *
 * Copyright (c) Coin Sciences Ltd
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */


	error_reporting(E_ALL | E_STRICT);
	
	require_once 'coinspark.php';
	
	$inputFileName=@$argv[1];

	if (file_exists($inputFileName)) {
		$inputSource=fopen($inputFileName, 'rb');
		ProcessInputContents($inputSource);
		fclose($inputSource);

	} else
		echo "No CoinSpark test input file was specified\n\n";
 

	function GetNextInputLine($inputSource)
	{
		$inputLine=fgets($inputSource, 65536);
		if (!strlen($inputLine)) // false or empty string both indicate end of file
			return null;
		
		$inputLine=rtrim($inputLine, "\n");

		$hashPos=strpos($inputLine, ' # ');
		if ($hashPos!==false)
			$inputLine=substr($inputLine, 0, $hashPos);
		
		return $inputLine;
	}
	
	
	function GetNextInputLines($inputSource, $countLines)
	{
		$inputLines=array();
		
		while ($countLines-->0) {
			$inputLine=GetNextInputLine($inputSource);
			if (!isset($inputLine))
				return null;
				
			$inputLines[]=$inputLine;
		}
		
		return $inputLines;
	}
	
	
	function ProcessInputContents($inputSource)
	{
		$header=GetNextInputLine($inputSource);
		GetNextInputLine($inputSource); // discard blank line

		switch ($header) {
			case 'CoinSpark Address Tests Input':
				ProcessAddressTests($inputSource);
				break;

			case 'CoinSpark AssetRef Tests Input':
				ProcessAssetRefTests($inputSource);
				break;
				
			case 'CoinSpark Script Tests Input':
				ProcessScriptTests($inputSource);
				break;
				
			case 'CoinSpark AssetHash Tests Input':
				ProcessAssetHashTests($inputSource);
				break;
				
			case 'CoinSpark Genesis Tests Input':
				ProcessGenesisTests($inputSource);
				break;
				
			case 'CoinSpark Transfer Tests Input':
				ProcessTransferTests($inputSource);
				break;
				
			case 'CoinSpark MessageHash Tests Input':
				ProcessMessageHashTests($inputSource);
				break;
		}
	}


	function ProcessAddressTests($inputSource)
	{
		echo "CoinSpark Address Tests Output\n\n";
		
		while (true) {
			$inputLine=GetNextInputLine($inputSource);
			if (!isset($inputLine))
				break;
				
			$address=new CoinSparkAddress();
			
			if ($address->decode($inputLine))
				echo $address->toString();
			else
				die("Failed to decode address: $inputLine\n");
				
			$encoded=$address->encode();
			
			if ($encoded!=$inputLine)
				die("Encode address mismatch: $encoded should be $inputLine\n");
				
			if (!$address->match($address))
				die("Failed to match address to itself!\n");
		}
	}


	function ProcessAssetRefTests($inputSource)
	{
		echo "CoinSpark AssetRef Tests Output\n\n";
	
		while (true) {
			$inputLine=GetNextInputLine($inputSource);
			if (!isset($inputLine))
				break;

			$assetRef=new CoinSparkAssetRef();
		
			if ($assetRef->decode($inputLine))
				echo $assetRef->toString();
			else
				die("Failed to decode AssetRef: $inputLine\n");
			
			$encoded=$assetRef->encode();
		
			if ($encoded!=$inputLine)
				die("Encode AssetRef mismatch: $encoded should be $inputLine\n");
				
			if (!$assetRef->match($assetRef))
				die("Failed to match assetRef to itself!\n");
		}
	}
	
	
	function ProcessScriptTests($inputSource)
	{
		echo "CoinSpark Script Tests Output\n\n";
		
		while (true) {
			$inputLines=GetNextInputLines($inputSource, 4);
			if (!isset($inputLines))
				break;
				
			list($countInputs, $countOutputs, $scriptPubKeyHex)=$inputLines;

			$metadata=CoinSparkScriptToMetadata($scriptPubKeyHex, true);
			if (!isset($metadata))
				die("Could not decode script metadata: $scriptPubKeyHex\n");
				
		//	Read in the different types of metadata
				
			$genesis=new CoinSparkGenesis();
			$hasGenesis=$genesis->decode($metadata);
			
			$paymentRef=new CoinSparkPaymentRef();
			$hasPaymentRef=$paymentRef->decode($metadata);
			
			$transfers=new CoinSparkTransferList();
			$hasTransfers=$transfers->decode($metadata, $countInputs, $countOutputs);
			
			$message=new CoinSparkMessage();
			$hasMessage=$message->decode($metadata, $countOutputs);
		
		//	Output the toString()s
			
			if ($hasGenesis)
				echo $genesis->toString();
			
			if ($hasPaymentRef)
				echo $paymentRef->toString();
				
			if ($hasTransfers)
				echo $transfers->toString();
				
			if ($hasMessage)
				echo $message->toString();
		
		//	Re-encode
		
			$testMetadata='';
			$testMetadataMaxLen=strlen($metadata);
			$nextMetadataMaxLen=$testMetadataMaxLen;
			
			$encodeOrder=array('genesis', 'paymentRef', 'transfers', 'message');
			
			foreach ($encodeOrder as $encodeField) {
				$triedNextMetadata=false;
				
				switch ($encodeField) {
					case 'genesis':
						if ($hasGenesis) {
							$nextMetadata=$genesis->encode($nextMetadataMaxLen);
							$triedNextMetadata=true;
						}
						break;
					
					case 'paymentRef':
						if ($hasPaymentRef) {
							$nextMetadata=$paymentRef->encode($nextMetadataMaxLen);
							$triedNextMetadata=true;
						}
						break;
					
					case 'transfers':
						if ($hasTransfers) {
							$nextMetadata=$transfers->encode($countInputs, $countOutputs, $nextMetadataMaxLen);
							$triedNextMetadata=true;
						}
						break;
						
					case 'message':
						if ($hasMessage) {
							$nextMetadata=$message->encode($countOutputs, $nextMetadataMaxLen);
							$triedNextMetadata=true;
						}
						break;
				}
				
				if ($triedNextMetadata) {
					if (!isset($nextMetadata))
						die("Failed to reencode $encodeField metadata!\n");

					if (strlen($testMetadata)) {
						$testMetadata=CoinSparkMetadataAppend($testMetadata, $testMetadataMaxLen, $nextMetadata);
						if (!isset($testMetadata))
							die("Insufficient space to append $encodeField metadata!\n");

					} else
						$testMetadata=$nextMetadata;
					
					$nextMetadataMaxLen=CoinSparkMetadataMaxAppendLen($testMetadata, $testMetadataMaxLen);
				}
			}
			
		//	Test other library functions while we are here
		
			if ($hasGenesis) {
				if (!$genesis->match($genesis, true))
					die("Failed to match genesis to itself!\n");
				
				if ($genesis->calcHashLen(strlen($metadata))!=$genesis->assetHashLen) // assumes that metadata only contains genesis
					die("Failed to calculate matching asset hash length!\n");
				
				$testGenesis=new CoinSparkGenesis();
				$testGenesis->decode($metadata);
				
				$rounding=rand(0, 2)-1;
				
				$testGenesis->setQty(0, 0);
				$testGenesis->setQty($genesis->getQty(), $rounding);
				
				$testGenesis->setChargeFlat(0, 0);
				$testGenesis->setChargeFlat($genesis->getChargeFlat(), $rounding);
				
				if (!$genesis->match($testGenesis, false))
					die("Mismatch on genesis rounding!\n");
			}
			
			if ($hasPaymentRef)
				if (!$paymentRef->match($paymentRef))
					die("Failed to match paymentRef to itself!\n");
			
			if ($hasTransfers) {
				if (!$transfers->match($transfers, true))
					die("Failed to strictly match transfers to itself!\n");

				if (!$transfers->match($transfers, false))
					die("Failed to leniently match transfers to itself!\n");
			}
			
			if ($hasMessage) {
				if (!$message->match($message, true))
					die("Failed to strictly match message to itself!\n");

				if (!$message->match($message, false))
					die("Failed to leniently match message to itself!\n");
					
				$messageEncode=$message->encode($countOutputs, strlen($metadata)); // encode on its own to check calcHashLen()
				
				if ($message->calcHashLen($countOutputs, strlen($messageEncode))!=$message->hashLen)
					die("Failed to calculate matching message hash length!\n");
			}
			
		//	Compare to the original
			
			$encoded=CoinSparkMetadataToScript($testMetadata, true);
			
			if ($encoded!=$scriptPubKeyHex)
				die("Encode metadata mismatch: $encoded should be $scriptPubKeyHex\n");
				
			$checkMetadata=CoinSparkScriptToMetadata(CoinSparkMetadataToScript($testMetadata, false), false);
			
			if ($checkMetadata!=$testMetadata)
				die("Binary metadata to/from script mismatch!\n");
		}
	}
	
	
	function ProcessAssetHashTests($inputSource)
	{
		echo "CoinSpark AssetHash Tests Output\n\n";
		
		while (true) {
			$inputLines=GetNextInputLines($inputSource, 10);
			if (!isset($inputLines))
				break;
				
			list($name, $issuer, $description, $units, $issueDate, $expiryDate, $interestRate, $multiple, $contractContent)=$inputLines;
		
			$hash=CoinSparkCalcAssetHash($name, $issuer, $description, $units, $issueDate, $expiryDate, $interestRate, $multiple, $contractContent);
			
			echo strtoupper(bin2hex($hash))."\n";
		}
	}
	
	
	function ProcessGenesisTests($inputSource)
	{
		echo "CoinSpark Genesis Tests Output\n\n";
		
		while (true) {
			$inputLines=GetNextInputLines($inputSource, 7);
			if (!isset($inputLines))
				break;
		
		//	Break apart and decode the input lines
			
			list($firstSpentTxId, $firstSpentVout, $metadataHex, $outputsSatoshisString, $outputsRegularString, $feeSatoshis)=$inputLines;

			$genesis=new CoinSparkGenesis();
			if (!$genesis->decode(pack('H*', $metadataHex)))
				die("Failed to decode genesis metadata: $metadataHex\n");
				
			$outputsSatoshis=explode(',', $outputsSatoshisString);
			$outputsRegular=explode(',', $outputsRegularString);			
			$countOutputs=count($outputsSatoshis);
		
			$validFeeSatoshis=$genesis->calcMinFee($outputsSatoshis, $outputsRegular);
				
		//	Perform the genesis calculation
			
			if ($feeSatoshis>=$validFeeSatoshis)
				$outputBalances=$genesis->apply($outputsRegular);
			else
				$outputBalances=array_fill(0, $countOutputs, 0);
				
		//	Output the results
			
			echo sprintf("%.0f", $validFeeSatoshis)." # transaction fee satoshis to be valid\n";
			foreach ($outputBalances as $outputIndex => $outputBalance)
				echo ($outputIndex ? ',' : '').sprintf("%.0f", $outputBalance);
			echo " # units of the asset in each output\n";
			echo $genesis->calcAssetURL($firstSpentTxId, $firstSpentVout)." # asset web page URL\n\n";
		}
	}

	
	function ProcessTransferTests($inputSource)
	{
		echo "CoinSpark Transfer Tests Output\n\n";
		
		while (true) {
			$inputLines=GetNextInputLines($inputSource, 8);
			if (!isset($inputLines))
				break;
		
		//	Break apart and decode the input lines
			
			list($genesisMetadataHex, $assetRefString, $transfersMetadataHex, $inputBalancesString,
				$outputsSatoshisString, $outputsRegularString, $feeSatoshis)=$inputLines;

			$genesis=new CoinSparkGenesis();
			if (!$genesis->decode(pack('H*', $genesisMetadataHex)))
				die("Failed to decode genesis metadata: $genesisMetadataHex\n");
				
			$assetRef=new CoinSparkAssetRef();
			if (!$assetRef->decode($assetRefString))
				die("Failed to decode asset reference: $assetRefString\n");
				
			$inputBalances=explode(',', $inputBalancesString);
			$outputsSatoshis=explode(',', $outputsSatoshisString);
			$outputsRegular=explode(',', $outputsRegularString);
			
			$countInputs=count($inputBalances);
			$countOutputs=count($outputsSatoshis);
		
			$transfers=new CoinSparkTransferList();
			if (!$transfers->decode(pack('H*', $transfersMetadataHex), $countInputs, $countOutputs))
				die("Failed to decode transfers metadata: $transfersMetadataHex\n");
			$validFeeSatoshis=$transfers->calcMinFee($countInputs, $outputsSatoshis, $outputsRegular);
				
		//	Perform the transfer calculation and get default flags
			
			if ($feeSatoshis>=$validFeeSatoshis)
				$outputBalances=$transfers->apply($assetRef, $genesis, $inputBalances, $outputsRegular);
			else
				$outputBalances=$transfers->applyNone($assetRef, $genesis, $inputBalances, $outputsRegular);
				
			$outputsDefault=$transfers->defaultOutputs($countInputs, $outputsRegular);
				
		//	Output the results
			
			echo sprintf("%.0f", $validFeeSatoshis)." # transaction fee satoshis to be valid\n";

			foreach ($outputBalances as $outputIndex => $outputBalance)
				echo ($outputIndex ? ',' : '').sprintf("%.0f", $outputBalance);
			echo " # units of this asset in each output\n";
			
			foreach ($outputsDefault as $outputIndex => $outputDefault)
				echo ($outputIndex ? ',' : '').($outputDefault ? '1' : '0');
			echo " # boolean flags whether each output is in a default route\n\n";
			
		//	Test the net and gross calculations using the input balances as example net values
		
			foreach ($inputBalances as $inputBalance) {
				$testGrossBalance=$genesis->calcGross($inputBalance);
				$testNetBalance=$genesis->calcNet($testGrossBalance);
				
				if ($inputBalance!=$testNetBalance)
					die(sprintf("Net to gross to net mismatch: %.0f -> %.0f -> %.0f!\n", $inputBalance, $testGrossBalance, $testNetBalance));
			}
		}
	}
	
	
	function ProcessMessageHashTests($inputSource)
	{
		echo "CoinSpark MessageHash Tests Output\n\n";
		
		while (true) {
			$inputLines=GetNextInputLines($inputSource, 2);
			if (!isset($inputLines))
				break;
				
			list($salt, $countParts)=$inputLines;
			
			$inputLines=GetNextInputLines($inputSource, 3*$countParts+1);
			if (!isset($inputLines))
				break;
				
			$messageParts=array();
			while ((count($messageParts)<$countParts) && count($inputLines))
				$messageParts[]=array(
					'mimeType' => array_shift($inputLines),
					'fileName' => array_shift($inputLines),
					'content' => array_shift($inputLines),
				);
			
			$hash=CoinSparkCalcMessageHash($salt, $messageParts);
			
			echo strtoupper(bin2hex($hash))."\n";
		}
	}